### PR TITLE
[Feature][WIP][Kernel] Kimi K3 DSpark three-operator precision investigation

### PR DESCRIPTION
**Pull request by maoxx241**
## What this PR does / why we need it

This is a **diagnostic companion PR** to #13277. It combines Kimi K3 MLA DSpark support with the `AttnResFwd`, `fused_norm_gate`, and NoPE `mla_preprocess` integrations so their target-model and speculative-decoding behavior can be compared directly.

The current head, `cdf6a4904`, replaces the draft-type-dependent attention-residual fallback with Kimi K3's canonical golden small-op decomposition for every execution mode. The AscendC `AttnResFwd` source remains in the diagnostic branch, but Kimi K3 no longer selects it from the model path while the numerical trajectory is being validated.

> [!IMPORTANT]
> The earlier conclusion that `torch_npu.npu_rms_norm` followed by FP32 `torch.matmul` was the Kimi K3 reference was incorrect. That decomposition is a vLLM-Ascend compatibility path introduced during the private Kimi integration, not the official Kimi K3 or upstream vLLM definition.

## Corrected root cause

The official Kimi K3 implementation and the NVIDIA, AMD, and SM100 vLLM implementations all use the same score formula:

```python
values = torch.cat((block_residual, prefix_sum.unsqueeze(1)), dim=1).float()
inverse_rms = torch.rsqrt(values.square().mean(-1, keepdim=True) + eps)
normalized_without_gamma = values * inverse_rms
score_weight = norm.weight.float() * projection.weight.squeeze(0).float()
scores = (normalized_without_gamma * score_weight).sum(-1)
probabilities = scores.softmax(-1).unsqueeze(1)
output = torch.matmul(probabilities, values).squeeze(1)
```

References:

- [Official Kimi K3 implementation](https://huggingface.co/moonshotai/Kimi-K3/blob/a527b42cb673d79569f3ffe0b3a0a655df98a739/modeling_kimi.py#L1119-L1132)
- [vLLM NVIDIA Triton implementation](https://github.com/vllm-project/vllm/blob/d9dac2b3d4cdd57ff2f7c5311bc8aaf3eef3feec/vllm/models/kimi_k3/nvidia/ops/attn_res.py#L102-L139)
- [vLLM AMD Triton implementation](https://github.com/vllm-project/vllm/blob/d9dac2b3d4cdd57ff2f7c5311bc8aaf3eef3feec/vllm/models/kimi_k3/amd/ops/attn_res.py#L41-L70)
- [vLLM SM100 CUDA implementation](https://github.com/vllm-project/vllm/blob/d9dac2b3d4cdd57ff2f7c5311bc8aaf3eef3feec/csrc/libtorch_stable/kimi_k3/attn_res_kernel.cu#L325-L349)

The original AscendC kernel also pre-combines `norm_weight * projection_weight` and therefore implements the canonical formula. The numerical baseline drift was introduced earlier by vLLM-Ascend commit `3d396f2f`, which replaced the canonical small ops with `torch_npu.npu_rms_norm(values, norm_weight)` followed by a separate FP32 projection matmul. Its CPU unit test mocked `npu_rms_norm`, so it verified routing and dtype behavior but not the real NPU reduction trajectory.

The AscendC output is not bitwise identical to the PyTorch golden because its hardware reductions are grouped differently, but it is closer to the canonical golden than to the legacy NPU decomposition:

| Shape | AscendC vs canonical golden mismatches | AscendC vs legacy NPU mismatches | Canonical vs legacy mismatches | Max BF16 error |
|---|---:|---:|---:|---:|
| `T=16`, blocks=8, H=7168 | 8 | 23 | 20 | 0.00390625 |
| `T=32`, blocks=8, H=7168 | 25 | 53 | 49 | 0.00390625 |
| `T=129`, blocks=8, H=7168 | 93 | 200 | 200 | 0.00390625 |

Therefore the earlier MLA acceptance change does not prove an operator correctness bug. It proves that the fixed draft checkpoint is sensitive to the target-model numerical trajectory:

| Target attention-residual trajectory | Accepted / draft steps | Mean acceptance length | Output token hash |
|---|---:|---:|---|
| Legacy `npu_rms_norm + matmul` | 830 / 196 | 5.23469387755102 | `e849aec1...12a` |
| Canonical AscendC | 762 / 265 | 3.8754716981132074 | `1f086e02...507` |

Acceptance length is a target/draft alignment metric, not a standalone target-model correctness oracle. The previous MLA-only fallback restored the legacy checkpoint alignment, but it made target semantics depend on which draft architecture was attached and was therefore withdrawn.

## Current implementation

Commit `cdf6a4904`:

- uses the canonical pre-combined golden small ops for attention, MLP, auxiliary-hidden-state, and output residual mixing;
- removes `use_fused_kernel` and all MLA/GQA draft-type routing from the model;
- removes the `torch_npu.npu_rms_norm + torch.matmul` compatibility decomposition;
- restores the operator golden to the official pre-combined formula; and
- retains `T=16` ACLGraph-padded decode coverage.

There is no semantic operation that runs matmul before RMSNorm. The earlier kernel experiment merely moved the gamma multiplication from the pre-combined score weight into the per-row path. Gamma and projection weights commute algebraically; separating them changes floating-point grouping but does not change the mathematical formula, and the experiment did not eliminate hardware reduction differences.

## Exact acceptance baselines

All rows use TP16 x DP4 x EP64, ACLGraph `FULL_DECODE_ONLY`, prefix cache off, DP rank 0, concurrency 1, exactly 8,192 prompt tokens and 1,024 completion tokens, greedy DSpark K=7, `temperature=0`, `seed=0`, and prompt-token SHA256 `4825bf16ae78f7b19384d8e6a356195a91aed115a8ddca743d921aa93b60f26b`.

The requested #13277 comparison baseline uses target `/mnt/weight/k3/Kimi-K3-w4a8-int-multicards` plus draft `/mnt/weight/k3/Kimi-K3-DSpark-GQA-20260801` (`epoch=10 step=1150`, model SHA256 `ecd746459b4a603ce0d2c64f73935efead29bd651b14439b99e57ee8b41b77ca`):

| Source | Accepted / draft tokens / draft steps | Mean AL | Per-token acceptance | Output hash | TTFT | TPOT | E2E |
|---|---:|---:|---:|---|---:|---:|---:|
| #13277 legacy target path | 868 / 1099 / 157 | 6.528662420382165 | 78.98089171974523% | `e849aec1...12a` | 3806.27 ms | 23.20 ms | 27.538 s |
| #13315 canonical golden small ops | 871 / 1071 / 153 | 6.69281045751634 | 81.32586367880486% | `3cd11a7a...745c` | 3954.91 ms | 23.87 ms | 28.372 s |

The canonical golden-small-op path therefore aligns with or exceeds the #13277 GQA checkpoint baseline on this fixed request: per-token acceptance is **+2.34497195905963 percentage points**, mean AL is **+0.164148037134175**, and the measured request needs four fewer draft steps. The output hash changes, but both rows return HTTP 200, exactly 1,024 completion tokens with `finish_reason=length`, and pass the benchmark quality checks. This single-request latency comparison is included for reproducibility, not as a throughput claim.

## How was this patch tested?

- `3 passed, 28 deselected`: focused Kimi K3 golden residual routing/math and auxiliary-stream tests in the real Ascend container environment.
- Direct A3 `T=16`, blocks=8, H=7168 smoke: exact equality with the canonical small-op golden, BF16 max error `0.0`.
- Ruff format/check and `git diff --check` pass for all changed files.
- Exact four-node 8K/1K GQA comparison: TP16 x DP4 x EP64, 64/64 workers, graph capture size 16, all four rank logs with zero fatal matches, HTTP 200, prompt/completion counts 8,192/1,024, prompt hash `4825bf16...f26b`, and `quality_passed=true`.
- Run provenance: the deployed source on all four nodes was guarded as `cdf6a4904d040537705f2ce67f46b541571e5071`; the unchanged #13277 benchmark harness still records its own `source_commit=cfaebdd6...` in the artifact manifest.

## Does this PR introduce any user-facing change?

Yes. Kimi K3 now uses one canonical attention-residual definition independent of whether the attached draft model is MLA, GQA, or absent.

## Test plan

- [x] Verify official Kimi K3 and upstream vLLM attention-residual semantics.
- [x] Switch the Kimi K3 model path to the canonical golden small ops.
- [x] Run focused unit tests and a realistic `T=16`, H=7168 A3 smoke.
- [x] Run the exact #13277 8K/1K case with multicard target plus the latest GQA-DSpark checkpoint.
- [x] Compare acceptance counters, output hash, and latency against the #13277 baseline above.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/d02df748bf9efd99022f1a062597dc3cb3808485

### CODE REVIEW

**Comment by maoxx241** [comment]
## Correction: canonical golden, operator semantics, and current implementation

The previous version of this comment treated the legacy vLLM-Ascend `torch_npu.npu_rms_norm + torch.matmul` path as the Kimi K3 model reference. Source comparison with the official Kimi K3 model, current upstream vLLM implementations, and the pre-optimization private Kimi history shows that conclusion was wrong.

### Canonical formula

Official Kimi K3 and upstream vLLM compute the residual score by normalizing without gamma and pre-combining the two learned weights:

```python
values = torch.cat((block_residual, prefix_sum.unsqueeze(1)), dim=1).float()
inverse_rms = torch.rsqrt(values.square().mean(-1, keepdim=True) + eps)
normalized_without_gamma = values * inverse_rms
score_weight = norm.weight.float() * projection.weight.squeeze(0).float()
scores = (normalized_without_gamma * score_weight).sum(-1)
probabilities = scores.softmax(-1).unsqueeze(1)
mixed = torch.matmul(probabilities, values).squeeze(1)
```

The original AscendC resident and reload kernels use this same mathematical grouping: they load/cast `normWeight` and `projWeight`, pre-compute `scoreWeight = normWeight * projWeight`, calculate the inverse RMS, and reduce `(values * inverse_rms) * scoreWeight`.

So the earlier request to "exchange RMSNorm and matmul" was based on a false reference. The attempted kernel change separated the two weights and moved gamma multiplication into the token loop:

```cpp
v = CastFp32(values);
invRms = CustomReduceRsqrt(v * v);
v = v * invRms;
v = v * normWeight;
score = ReduceSum(v * projWeight);
```

This is algebraically equivalent, but it changes floating-point grouping away from the official pre-combined golden. It did not remove the remaining reduction-order differences and was correctly reverted. No AscendC source modification is part of the current fix.

### Where the drift actually entered

The pre-optimization private Kimi implementation used the canonical pre-combined formula. vLLM-Ascend commit `3d396f2f` changed that model path to real `npu_rms_norm` followed by a separate FP32 projection matmul. That path became the #13277 acceptance baseline, but it is not the official/upstream golden.

Relabeling the real layer-12 A3 diagnostics gives:

| Shape | AscendC vs canonical golden | AscendC vs legacy NPU path | Canonical vs legacy |
|---|---:|---:|---:|
| `T=16`, blocks=8, H=7168 | 8 | 23 | 20 |
| `T=32`, blocks=8, H=7168 | 25 | 53 | 49 |
| `T=129`, blocks=8, H=7168 | 93 | 200 | 200 |

The maximum BF16 difference is `0.00390625`. AscendC is not bitwise identical to the PyTorch golden because its hardware reductions are grouped differently, but it is closer to the canonical implementation than to the legacy NPU path.

The earlier MLA result must therefore be interpreted as a target/draft trajectory change, not as proof that AscendC is incorrect:

| Target path | Accepted / draft steps | Mean AL | Output hash |
|---|---:|---:|---|
| Legacy `npu_rms_norm + matmul` | 830 / 196 | 5.23469387755102 | `e849aec1...12a` |
| Canonical AscendC | 762 / 265 | 3.8754716981132074 | `1f086e02...507` |

### Current PR change

Commit `cdf6a4904` switches all Kimi K3 attention-residual calls to the canonical golden small ops and removes both the fused/reference selector and the MLA/GQA draft-type-dependent routing. In particular:

- `vllm_ascend/ops/attn_res.py` now contains the pre-combined golden above;
- `vllm_ascend/models/kimi_k3.py` no longer passes `use_fused_kernel` or detects the attached draft type for residual semantics;
- the `npu_rms_norm + matmul` compatibility path is removed; and
- the focused golden test keeps the ACLGraph-padded `T=16` case.

Focused validation completed so far:

- `3 passed, 28 deselected` for Kimi K3 residual routing/math and auxiliary-stream tests;
- A3 `T=16`, blocks=8, H=7168 canonical small-op smoke is bitwise exact (`max_abs_error=0.0`);
- Ruff format/check and `git diff --check` pass.

### Requested GQA 8K/1K result

The exact rerun is complete with the same multicard target, latest `Kimi-K3-DSpark-GQA-20260801` draft, TP16 x DP4 x EP64 topology, prefix cache off, DP rank 0, concurrency 1, greedy K=7 sampling, and prompt-token hash `4825bf16...f26b`:

| Source | Accepted / draft tokens / draft steps | Mean AL | Per-token acceptance | Output token hash | TTFT | TPOT | E2E |
|---|---:|---:|---:|---|---:|---:|---:|
| #13277 legacy target path | 868 / 1099 / 157 | 6.528662420382165 | 78.98089171974523% | `e849aec1...12a` | 3806.27 ms | 23.20 ms | 27.538 s |
| #13315 canonical golden small ops | 871 / 1071 / 153 | 6.69281045751634 | 81.32586367880486% | `3cd11a7a...745c` | 3954.91 ms | 23.87 ms | 28.372 s |

The golden-small-op implementation is therefore **+2.34497195905963 percentage points** in per-token acceptance and **+0.164148037134175** in mean AL relative to #13277 on this fixed case; it also uses four fewer draft steps. Both measured requests return HTTP 200, exactly 8,192 prompt and 1,024 completion tokens, `finish_reason=length`, and `quality_passed=true`. The different output hash shows the target numerical trajectory changed, while the acceptance result confirms that the latest GQA draft remains at least as well aligned as the #13277 baseline.

Four-node audit: 64/64 workers started, graph capture size 16 completed on all four DP ranks, and all four rank logs have zero fatal matches. The deployed source was guarded as `cdf6a4904d040537705f2ce67f46b541571e5071` on every node. The unchanged benchmark harness is copied from the #13277 reproduction and consequently keeps `source_commit=cfaebdd6...` as harness provenance in its manifest; it is not the deployed service commit.
